### PR TITLE
Initial device reporting in config daemon

### PR DESCRIPTION
**Pull request by almaslennikov**

### CODE REVIEW

File: pkg/host/utils.go
Comment (ykulazhenkov) [inline pkg/host/utils.go:19]: Can you add comments to the Interface and functions?

File: pkg/daemon/daemon.go
Comment (ykulazhenkov) [inline pkg/daemon/daemon.go:16]: Can you add comments to the struct and functions?

**Review by ykulazhenkov — COMMENTED** [review]
partial review

File: config/daemon/daemon.yaml
Comment (adrianchiris) [inline config/daemon/daemon.yaml:25]: lets keep original value here.

Comment (adrianchiris) [reply config/daemon/daemon.yaml:25]: this customizations should be made with kustomize when deploying

File: config/daemon/daemon.yaml
Comment (adrianchiris) [inline config/daemon/daemon.yaml:26]: remove altogether if its not needed.

File: go.mod
Comment (adrianchiris) [reply go.mod:10]: +1

File: go.mod
Comment (adrianchiris) [inline go.mod:7]: should we just shamelessly copy the dependencies ? we dont claim to have stable API for pkgs in device plugin.

File: pkg/log/log.go
Comment (adrianchiris) [inline pkg/log/log.go:1]: usually pkg name matches the folder.

why not call this one log ?

Comment (adrianchiris) [reply pkg/log/log.go:1]: also do you need this pkg ?

main reason to have it is to support runtime updates of log level.

that said its fine to have it for intialization of zap logger.

File: cmd/nic-configuration-daemon/main.go
Comment (adrianchiris) [inline cmd/nic-configuration-daemon/main.go:50]: what about signal handling ?

File: cmd/nic-configuration-daemon/main.go
Comment (adrianchiris) [inline cmd/nic-configuration-daemon/main.go:48]: daemon is going to have several parts ("controllers")
what do you think about using controller manager for it ?

you can register reconcilers as well as add "runnable" components if needed.

File: pkg/daemon/daemon.go
Comment (adrianchiris) [inline pkg/daemon/daemon.go:59]: so its one shot ? shouldnt it run periodically ?

Comment (adrianchiris) [reply pkg/daemon/daemon.go:59]: what if error happens on one device ? when would it be retried ?

File: pkg/daemon/daemon.go
Comment (adrianchiris) [inline pkg/daemon/daemon.go:43]: this should be the operator namespace no ?

File: pkg/daemon/daemon.go
Comment (adrianchiris) [inline pkg/daemon/daemon.go:64]: we do want some valid context here : e.g a context that is cancelled if signal recieved.

File: pkg/host/utils.go
Comment (adrianchiris) [inline pkg/host/utils.go:1]: please add license headers for everything

File: pkg/host/utils.go
Comment (adrianchiris) [inline pkg/host/utils.go:68]: do we want "ToLower" for these strings ?

File: pkg/host/utils.go
Comment (adrianchiris) [inline pkg/host/utils.go:82]: what if one found but not the other ?

File: pkg/host/utils.go
Comment (adrianchiris) [inline pkg/host/utils.go:104]: same WRT the "ToLower" comment

File: pkg/host/utils.go
Comment (adrianchiris) [inline pkg/host/utils.go:114]: nit: ensure both are non-empty

File: pkg/host/utils.go
Comment (adrianchiris) [inline pkg/host/utils.go:160]: can we leverage rdmamap lib for it ?

https://github.com/Mellanox/rdmamap/blob/d8230af2dc589a7e0a421e7b9927284d838d878a/rdma_map.go#L386

File: pkg/host/host.go
Comment (adrianchiris) [inline pkg/host/host.go:87]: there is a lib in api machinery to deal with conditions. i strongly suggest to use it.

as an example. you must set last transition time, which it will take care for u.
moreover if there are other conditions this function will remove them and replace with this single condition.

**Review by adrianchiris — COMMENTED** [review]
what are the plans for unit-tests ?

File: go.mod
Comment (almaslennikov) [reply go.mod]: Changed it to a recently released v1.3.0

File: pkg/consts/consts.go
Comment (almaslennikov) [reply pkg/consts/consts.go]: Done for all consts

File: go.mod
Comment (almaslennikov) [reply go.mod]: Copied the function locally

File: pkg/log/log.go
Comment (almaslennikov) [reply pkg/log/log.go]: Changed the folder name to match the package

File: cmd/nic-configuration-daemon/main.go
Comment (almaslennikov) [reply cmd/nic-configuration-daemon/main.go]: Changed daemon to be a NicDevice reconciler, now it uses a controller manager

File: pkg/daemon/daemon.go
Comment (almaslennikov) [reply pkg/daemon/daemon.go]: Resolved by using controller manager

File: pkg/host/utils.go
Comment (almaslennikov) [reply pkg/host/utils.go:102]: Now it returns an error

File: pkg/host/utils.go
Comment (almaslennikov) [reply pkg/host/utils.go:204]: Done

File: pkg/host/host.go
Comment (almaslennikov) [reply pkg/host/host.go]: Done, the logic will evolve, now added only a basic condition

File: Dockerfile.nic-configuration-daemon
Comment (almaslennikov) [reply Dockerfile.nic-configuration-daemon:29]: Not really, is there a reason not to?

**Comment by almaslennikov** [comment]
@adrianchiris @ykulazhenkov addressed your comments for the current review iteration.
the unit tests are to follow

File: Dockerfile.nic-configuration-daemon
Comment (ykulazhenkov) [reply Dockerfile.nic-configuration-daemon:29]: I mean currently you don't install anything to the image and I don't see why you need to switch from the distrolless. Do we need any OS packages in the container?

File: pkg/host/host.go
Comment (almaslennikov) [reply pkg/host/host.go:43]: We are executing a lot of system tools, where it's not always entirely clear, which system directories they try to access. If needed, we can try to make do with a handful of specific directories mounts but that's subject to problems in the future IMHO

File: internal/controller/nicdevice_controller.go
Comment (almaslennikov) [reply internal/controller/nicdevice_controller.go]: Changed the design to have two runnables as part of controller manager:
1. Device registry which is a simple runnable executing every 5 minutes. It discovers the devices and creates/deletes/updates the status if needed
2. NicDevice CR reconciler. It's triggered by spec updates and will apply the requested configuration and conditions

File: internal/controller/nicdevice_controller.go
Comment (almaslennikov) [reply internal/controller/nicdevice_controller.go]: Good idea, thanks

File: internal/controller/nicdevice_controller.go
Comment (almaslennikov) [reply internal/controller/nicdevice_controller.go]: See comment above

File: pkg/host/utils.go
Comment (almaslennikov) [reply pkg/host/utils.go]: changed to start with lower case

File: internal/controller/nicdevice_controller.go
Comment (ykulazhenkov) [inline internal/controller/nicdevice_controller.go:75]: Should you call the status update only in case if the status field was modified?

File: internal/controller/nicdevice_controller.go
Comment (ykulazhenkov) [inline internal/controller/nicdevice_controller.go:62]: Currently, If the spec is empty the status will not be updated with the "spec not set" condition. Is this intentional?

File: pkg/host/host.go
Comment (ykulazhenkov) [reply pkg/host/host.go:44]: If I understand correctly, this will case that we will call `mstconf` and `mstvpd` from the host (meaning these tools should be instaleld on the host OS) . I think this is not what we want.

File: config/daemon/daemon.yaml
Comment (almaslennikov) [reply config/daemon/daemon.yaml]: No, we don't. Removed this

File: internal/controller/nicdevice_controller.go
Comment (almaslennikov) [reply internal/controller/nicdevice_controller.go:153]: Reworked the status conditions. No they are only set in the controller (apart from the initial condition)

File: Dockerfile.nic-configuration-daemon
Comment (almaslennikov) [reply Dockerfile.nic-configuration-daemon:29]: Let's use distroless for now and switch later if needed

File: pkg/host/host.go
Comment (almaslennikov) [reply pkg/host/host.go:40]: @ykulazhenkov Even if we install those tools inside the container, they will still need to access different system directories, and we don't know fully which ones

Comment (adrianchiris) [reply pkg/host/host.go:40]: this should be addressed on a ad-hoc basis (if additional mounts needed)
for now we should try to have those tool part of the container image invoking them from the container,

e,g for mstflint related tools we do so today in sriov network config daemon

File: pkg/registry/suite_test.go
Comment (adrianchiris) [inline pkg/registry/suite_test.go:70]: nit: you use k8s 1.31 in go.mod 

should you use envtest of 1.31 or 1.30 if envtest has issues with 1.31?

File: pkg/host/host.go
Comment (almaslennikov) [reply pkg/host/host.go]: Now we install mstflint to our docker image, and only mount /sys dir (strace on all the used commands didn't show any additional dirs that we might need to mount) so far. That said, using the centos image as a base now

File: pkg/registry/deviceregistry.go
Comment (almaslennikov) [reply pkg/registry/deviceregistry.go]: Let's add this, it adds clarity for the user when listing CRs

File: pkg/registry/deviceregistry_test.go
Comment (almaslennikov) [reply pkg/registry/deviceregistry_test.go]: Changed the approach a little to use controller manager for starting the device discovery, it allows us to test that indexing by node name works as expected. Otherwise left tests as they were before

File: pkg/registry/deviceregistry.go
Comment (almaslennikov) [reply pkg/registry/deviceregistry.go]: According to [this](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications) we can't set a node as owner for our resources

Comment (adrianchiris) [reply pkg/registry/deviceregistry.go]: here is what i see in the link you sent:

> Cross-namespace owner references are disallowed by design. **Namespaced dependents can specify cluster-scoped** or namespaced owners.

Comment (adrianchiris) [reply pkg/registry/deviceregistry.go]: if you prefer not to set owner Ref which will delete NicDevice if node deleted. then please add logic to handle the following case:
- Node removed from k8s
- NICDevices associated with that node should be garbage collected.

Comment (almaslennikov) [reply pkg/registry/deviceregistry.go]: I see now that there is similar logic in Maintenance operator. I'll verify that it works and if so would prefer the NicDevices to be cleaned up automatically :)

File: Dockerfile.nic-configuration-daemon
Comment (adrianchiris) [inline Dockerfile.nic-configuration-daemon:27]: i dont think we care about s390x arch.

File: internal/controller/devicediscovery_controller.go
Comment (adrianchiris) [inline internal/controller/devicediscovery_controller.go:83]: One thing im worried about here: 
in DiscoverNicDevices its enough for any command to fail and the device is skipped thus not present in observedDevices which will trigger deletion of NICDevices (loosing their condition, state etc).

perhaps we should be strict in DiscoverNicDevices ? that is:

1. if we fail to list PCI devices -> we fail and retry
2. if we fail to get any information on a mlnx device -> we fail and retry
3. retry in this case should be sooner than 5 min imo.

WDYT ?

Comment (adrianchiris) [reply internal/controller/devicediscovery_controller.go:83]: with current logic, device will disappear for at least 5 minutes, then get recreated and reconfigured. (loosing its last applied state)

Comment (almaslennikov) [reply internal/controller/devicediscovery_controller.go:83]: One thing to notice is that if we make DiscoverNicDevices stricter, one faulty device will block all devices on that node from creating if we consistently fail to get some info about it, which could be okay as long as we report it properly.

WDYT?

Comment (almaslennikov) [reply internal/controller/devicediscovery_controller.go:116]: So I would prefer making DiscoverNicDevices a bit stricter and fail if we can't list PCI devices. I wouldn't fail when we can't get some device information on Mellanox devices. For example, there can be management interfaces like this:
`3b:00.2 DMA controller [0801]: Mellanox Technologies MT42822 BlueField-2 SoC Management Interface [15b3:c2d3]`
That do not support mstvpd calls. So we either make a list of device types that we want to skip (there can be others) or not fail on PN/SN/FW version/PSID retrievals.

When we do fail, though, we should requeue right away

Comment (adrianchiris) [reply internal/controller/devicediscovery_controller.go:116]: discovery logic should skip non network PCI device class with IB or ETH subclass see [1]
we do the this in sriov device plugin.

for what is left id expect this to work. anyway i leave it up to you to decide how to handle.

[1] https://pcisig.com/sites/default/files/files/PCI_Code-ID_r_1_11__v24_Jan_2019.pdf

Comment (almaslennikov) [reply internal/controller/devicediscovery_controller.go:116]: Makes sense to me. Added filtering based on the PCI device class, and now we fail when some properties of mellanox network device couldn't be retrieved.

**Comment by almaslennikov** [comment]
@adrianchiris addressed all your comments, PTAFL

File: internal/controller/devicediscovery_controller.go
Comment (adrianchiris) [inline internal/controller/devicediscovery_controller.go:165]: how can this happen ?

Comment (almaslennikov) [reply internal/controller/devicediscovery_controller.go:165]: If the device gets created but subsequent status update fails. We can't combine those two in the same API call. It's really a corner case but still can happen

File: internal/controller/devicediscovery_controller.go
Comment (almaslennikov) [reply internal/controller/devicediscovery_controller.go:133]: Fixed and added a unit test to check owner references

File: internal/controller/devicediscovery_controller.go
Comment (adrianchiris) [reply internal/controller/devicediscovery_controller.go:141]: that means device serial number is empty -> it will get deleted  at L97

up2u if you want to keep this logic, im probably missing something here.

File: internal/controller/devicediscovery_controller.go
Comment (adrianchiris) [inline internal/controller/devicediscovery_controller.go:132]: if this can fail then i would check for it.

**Review by adrianchiris — APPROVED** [review]
final round of comments. 

LGTM, once addressed feel free to self-merge.

File: internal/controller/devicediscovery_controller.go
Comment (almaslennikov) [reply internal/controller/devicediscovery_controller.go:141]: Not really. We list devices based on their `node` field in the status. If the status is empty, the cr will not match the selector. Thus, we try to recover a device without status

**Pull request merged**
